kvserver,logstore: move log appends to logstore

pav-kv opened the pull request:
This change moves a few components related to storing the newly appended Raft entries from `kvserver` to the `logstore` package.

Part of #91979
Release note: None

pav-kv committed:
logstore: add log-specific Ready struct

Release note: None

pav-kv committed:
kvserver,logstore: move RaftState struct to logstore

Release note: None

pav-kv committed:
kvserver,logstore: move logAppend to logstore

Release note: None

pav-kv committed:
kvserver: factor out logAppend stats

Release note: None

pav-kv committed:
kvserver: tidy up and pre-export logAppendStats

Release note: None

pav-kv committed:
kvserver,logstore: move append stats to logstore

Release note: None

pav-kv committed:
kvserver,logstore: move LogStore type to logstore

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/91951)
<!-- Reviewable:end -->

pav-kv committed:
logstore: unexport logAppend helper

Release note: None

pav-kv committed:
kvserver,logstore: move sideload purging to logstore

This commit moves the piece of code that purges sideloaded storage to logstore
package, and unexports the corresponding sideload helper. Probably other bits
using the sideload storage should move and be unexported too, because
SideloadStorage is effectively an implementation detail / optimization in the
log storage.

Release note: None

tbg reviewed (approved):
![image](https://user-images.githubusercontent.com/5076964/202476400-20f6478a-0272-4787-8f23-927bce446a20.png)

pavelkalinnikov commented on pkg/kv/kvserver/logstore/logstore.go:
```diff
@@ -127,3 +142,94 @@ type AppendStats struct {
+	// Synchronously commit the batch with the Raft log entries and Raft hard
+	// state as we're promising not to lose this data.
+	//
+	// Note that the data is visible to other goroutines before it is synced to
```
The commit that introduced this comment: 82cbb49f.

pavelkalinnikov replied:
Not sure if it was a drive-by comment at the time.

pavelkalinnikov replied:
I added a TODO to revisit this comment later. It seems to be in a wrong place now, `logserver` is now less responsible for maintaining Raft invariants.

pav-kv committed:
logstore: add TODO to revisit the old comment

Release note: None

pavelkalinnikov commented:
bors r=tbg
